### DISCUSSION TITLE
os.popen leaving a zombie

### DISCUSSION THREAD

**Discussion by protectivedad** [micropython/micropython#13239] [Unix & Windows] [answered]
I'm not very familiar with python so this might be just a simple problem. I am running micropython 1.18 on openwrt 22 on a wall switch (rt5350) and can't easily upgrade. I am using micropython to communicate with a mqtt broker. It is working for the most part, but one of the things it is doing is checking the status of the uhttpd server (and controlling it). The code for checkoing the status is:
```
def publish_https():
        with os.popen("service uhttpd status") as s:
                r = s.read()

        if r.strip() == "running":
                c.publish(HTTPS_STAT, "ON")
                https_signal.on()
        elif r.strip() == "inactive":
                c.publish(HTTPS_STAT, "OFF")
                https_signal.off()
        else:
                c.publish(HTTPS_STAT, "UNKNOWN")
                https_signal.on()
```

So I run a command check the output and publish the result. The problem is that the command leaves a zombie process everytime it is run.
```
 3560 root      3052 S    {mqtt-client.py} /usr/bin/micropython /etc/mosquitto/mqtt-client.py
 3561 root         0 Z    [mqtt-client.py]
```
When I start I publish the current status and this is what I get. What am I doing wrong?

**Comment by peterhinch** [comment]
I'm not sure what you're running here. MicroPython doesn't support pipes or subprocesses. Note the lack of `popen`:
```python
MicroPython v1.20.0-450-g3637252b7 on 2023-09-06; linux [GCC 11.4.0] version
Use Ctrl-D to exit, Ctrl-E for paste mode
>>> import os
>>> dir(os)
['__class__', '__name__', 'remove', 'sep', '__dict__', 'VfsFat', 'VfsLfs1', 'VfsLfs2', 'VfsPosix', 'chdir', 'errno', 'getcwd', 'getenv', 'ilistdir', 'listdir', 'mkdir', 'mount', 'putenv', 'rename', 'rmdir', 'stat', 'statvfs', 'system', 'umount', 'unlink', 'unsetenv', 'urandom']
```
or
```python
>>> 'popen' in dir(os)
False
```

**Comment by protectivedad** [comment]
It's supported in the [micropython-lib](https://github.com/micropython/micropython-lib/blob/7cdf70881519c73667efbc4a61a04d9c1a49babb/unix-ffi/os/os/__init__.py#L297). Maybe I posted this in the wrong forum. I didn't know there was one for discussing the libraries, I figured they were so closely connected that people here would know. Sorry about that. Any idea where I should ask my question?

**Reply by peterhinch** [reply]
You are asking in the right place. I'm not familiar with the unix-ffi libraries. Hopefully someone else will respond.

**Comment by dlech** [comment]
`popen` is calling `fork` which creates a child process. In order to avoid a zombie process, the parent process must wait on the child process. But since `popen` doesn't return the `pid` of the child, there is not currently a way to wait on the child process. So this looks like a bug in micropython-lib.

**Comment by protectivedad** [comment, marked as answer]
> `popen` is calling `fork` which creates a child process. In order to avoid a zombie process, the parent process must wait on the child process. But since `popen` doesn't return the `pid` of the child, there is not currently a way to wait on the child process. So this looks like a bug in micropython-lib.

Should I open an issue do you think? Is it fixable? The zombies make the function pretty useless, for me at least.

I needed the output so I used:
```
STATUS = "/tmp/uhttpd.status"

def publish_https():
        e = os.system("service uhttpd status >" + STATUS)
        with open(STATUS, "r") as f:
                r = f.read().strip()
        os.remove(STATUS)

        if r == "running":
                c.publish(HTTPS_STAT, "ON")
                https_signal.on()
        elif r == "inactive":
                c.publish(HTTPS_STAT, "OFF")
                https_signal.off()
        else:
                c.publish(HTTPS_STAT, "UNKNOWN")
                https_signal.on()
```

**Reply by dlech** [reply]
I think it would be difficult to fix it completely purely in Python since MicroPython currently doesn't allow implementing finalizers in Python (we need a way to ensure the PID is always waited for no matter what). But it certainly could be improved in micropython-lib.

**Reply by protectivedad** [reply]
Thanks. I opened and [issue](https://github.com/micropython/micropython-lib/issues/780) and will use the work around for now. Thanks for the time.
